walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,118 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
 
+/**
+ * This class only contains the query used to find the time ranges when a
+ * meeting request coincides with available time.
+ */
 public final class FindMeetingQuery {
+  // This comparator is used to sort a list of events by start time
+  // in ascending order.
+  private Comparator<Event> eventComparator = new Comparator<Event>() {
+    @Override
+    public int compare(Event event1, Event event2) {
+      return event1.getWhen().start() - event2.getWhen().start();
+    }
+  };
+  /**
+   * The way the query works is by going through the sorted events, and
+   * if the current event does not have any of the people in the meeting
+   * request, the event is skipped. If it does share people with the request,
+   * availableTimeDuration is set to the available time from the end of the
+   * previous event until the start of the current event. Then, if
+   * availableTimeDuration is at least the same size as the duration of the
+   * request, a TimeRange from pastEventEnd to currentEventStart is
+   * added to the list of available times.
+   *
+   * Time complexity: 
+   * n: number of events
+   * m: number of attendees in every event
+   * If m > logn, then O(nm)
+   * Otherwise, O(nlogn)
+   *
+   * Space complexity:
+   * n: number of events
+   * m: number of available time ranges
+   * O(m+n)
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> eventList = new ArrayList(events);
+    // Here eventList is sorted by event start time in ascending order.
+    Collections.sort(eventList, eventComparator);
+    // Since there are no previous events yet, the start of day is also the
+    // start of the first range to check available time.
+    int pastEventEnd = TimeRange.START_OF_DAY;
+    int currentEventStart;
+    int availableTimeDuration;
+    List<TimeRange> availableTimes = new ArrayList<>();
+    for (Event event : eventList) {
+      // If the event does not share attendes with the request, it means that
+      // it does not change time availability, so it's skipped.
+      if (!eventSharesAttendeesWithRequest(event, request)) continue;
+      // Available time goes from the end of the previous event to the start of
+      // the current event.
+      currentEventStart = event.getWhen().start();
+      availableTimeDuration = currentEventStart - pastEventEnd;
+      // This case is if the current event starts after the previous one ends.
+      // Example: |--A--|   |--B--|
```

Comment:
nit: put comments within the blocks that they are referring. So for instance, this would be:

```
if (currentEventStart > pastEventEnd) {
     // This case is if the current event starts after the previous one ends.
     // Example: |--A--|   |--B--|
      if (availableTimeDuration >= request.getDuration()) {
          availableTimes.add(TimeRange.fromStartDuration(
              pastEventEnd, availableTimeDuration));
      }
      // Regadrdless of whether the duration was enough for the meeting
     // the next available time will start when the current event ends.
     pastEventEnd = event.getWhen().end();
} 
```

Reply:
Done